WIP Tari script

stringhandler opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
WIP PR to track changes from tari script branch to development

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [ ] I'm merging against the `development` branch.
* [ ] I have squashed my commits into a single commit.

Commit:
Add tari script transaction data structures

Added tari script transaction data structures as per RFC-0201
<https://rfc.tari.com/RFC-0201_TariScript.html>, used default
tari script values everywhere. Also updated comms wire mode
to keep this network seperate.
TODO:
- Fix all unit tests
- Migrate tari script into all transaction protocols

wip

Commit:
Add tari script transaction data structures

Added tari script transaction data structures as per RFC-0201
<https://rfc.tari.com/RFC-0201_TariScript.html>, used default
tari script values everywhere. Also updated comms wire mode
to keep this network seperate.
TODO:
- Fix all unit tests
- Migrate tari script into all transaction protocols

wip

Commit:
Updated `/// Sum of script offsets for all kernels in this block.`

Commit:
Merge #2758: [common] Add tari script transaction data structures

Add tari script transaction data structures
Updated `/// Sum of script offsets for all kernels in this block.`

philipr-za committed:
Implement TariScript base node validation and default scripts in transactions

This PR includes the work of @SWvheerden, @hansieodendaal and myself to implement the foundations of TariScript support in the Tari base node and wallets.

The goal of this PR is to add all the data required for TariScripts while building transactions and also to validate the TariScripts by base nodes. 

On the wallet side of things this PR updates the wallet Sql database to store all the elements that the wallet needs to remember to spend an output with a script and updates the Transaction Negotiation Protocols to work with TariScripts. The updates are able to support general TariScripts but in the Wallet the current assumption is that all outputs will have the “Default” Nop script for now.

This PR also adds the fields to the headers required for script validation and the base node validation has been updated to also check the script elements for individual transactions and the accumulated offset for  constructed blocks and the blockchain.

One last feature that is included in this PR as it was easy to add in during the Transaction Protocol refactor is that the Coin Split function in the Wallet now produces outputs that are rewindable and thus can be recovered by a wallet recovery.

SWvheerden committed:
Updated `/// Sum of script offsets for all kernels in this block.`

Implement TariScript base node validation and default scripts in transactions

This PR includes the work of @SWvheerden, @hansieodendaal and myself to implement the foundations of TariScript support in the Tari base node and wallets.

The goal of this PR is to add all the data required for TariScripts while building transactions and also to validate the TariScripts by base nodes. 

On the wallet side of things this PR updates the wallet Sql database to store all the elements that the wallet needs to remember to spend an output with a script and updates the Transaction Negotiation Protocols to work with TariScripts. The updates are able to support general TariScripts but in the Wallet the current assumption is that all outputs will have the “Default” Nop script for now.

This PR also adds the fields to the headers required for script validation and the base node validation has been updated to also check the script elements for individual transactions and the accumulated offset for  constructed blocks and the blockchain.

One last feature that is included in this PR as it was easy to add in during the Transaction Protocol refactor is that the Coin Split function in the Wallet now produces outputs that are rewindable and thus can be recovered by a wallet recovery.

fix unit tests

Some code review comments

Commit:
Merge #2794: [common] Implement TariScript base node validation and default scripts in transactions

Implement TariScript base node validation and default scripts in transactions

Commit:
Fix sqlite3 database migrations for tari-script

- Fixed error "Cannot add a NOT NULL column with default value NULL" with
  `outputs` table; cannot add a `NOT NULL` column after the table has been
  created.
- Date integrity: Changed `commitment` and `hash` default `NULL` columns in
  `outputs` table to `NOT NULL`.
- Superfluous constraints: Changed all `NULL DEFAULT NULL` coulmns to `NULL`
  because `NULL` columns are `NULL` by default.

Commit:
Add send one-sided tx to console wallet TUI

Added the one-sided payment transaction sending functionality to
the console wallet TUI.
TODO: Receiver wallet scanning the blockchain for one-sided
      transactions.

Commit:
Add one-sided payment send transaction

Added one-sided to other payment transaction to the
transaction sending protocols. Consecutive PRs must
extend the wiring into the user interfaces. Added
unit tests.

SWvheerden committed:
Merge branch 'tari-script' into sw_tar-Script_validation

Commit:
Merge #2833: [chore] Remove the text_message_service from the wallet (#2833)

Commit:
Fix compile errors

Commit:
Merge #2850: [common] Fix compile errors

Fix compile errors

Commit:
Merge #2838: [wallet] Fix sqlite3 database migrations for tari-script

Fix sqlite3 database migrations for tari-script

SWvheerden committed:
fix cucumber tests to use standalone miner

fix build of wallet_ffi

Update integration_tests/features/Propagation.feature

Co-authored-by: Mike the Tike <stringhandler@gmail.com>
update name 

fix naming

Commit:
Merge #2835: [tests] Fix cucumber tests to use standalone miner

fix cucumber tests to use standalone miner

SWvheerden committed:
sw_squash tari_script

Commit:
Merge #2858: [wallet] Add one-sided payment send transaction

Add one-sided payment send transaction

Commit:
Add one sided tx to the command line

Added one sided transaction sending to the command line
interface.

philipr-za committed:
[Wallet] Update wallet gRPC server to send one-sided payments

This PR adds an enum to the PaymentRecipient Protocol struct that is used with the Transfer method to allow the user to specify if a payment is made using the traditional Mimblewimble method or as a One-sided payment.

This was tested manually as the Cucumber tests seem broken in the TariScript branch currently.

SWvheerden committed:
Merge branch 'tari-script' into sw_tari_script

Commit:
Rebase

libwallet-0.16.25
[Wallet] Persist transaction confirmations for complete transaction and expose via FFI
[base-node] Add rpc sessions to base node status line
[tests] Attempt to replicate sync issue in integration tests
Add wallet recovery cucumber test, seed words to file
Add reorg tests .GT. `orphan_storage_capacity`
Improve wallet recovery resiliency
Fix console wallet menu spacing
[base-node] Always output message when base node exits with an error
Add database concurrency check
Fix README installation artefacts path
Fixed cucumber tests
v0.8.8
[comms] Use grace period on client side handshake
Reduce checklist on PR template
[common] Fix key manager panic
[wallet] Leverage streaming for UTXO recovery
[Wallet] Fix wallet service API loop blocking
[chore] Fix clippy errors
sw_squash tari_script

Commit:
Merge #2866: [wallet] Add one sided tx to the command line

Commit:
Merge #2863: [wallet] Add send one-sided tx to console wallet TUI

Add send one-sided tx to console wallet TUI

Commit:
Merge #2865: [wallet] Update wallet gRPC server to send one-sided payments

the commit.

philipr-za committed:
Fix Cucumber tests in `TariScript` branch

This PR aims to fix the broken cucumber tests in the TariScript branch. The PR contains the following:

- Implement the TariScript logic for producing coin bases and transactions in the cucumber support javascript. The updates add a no_op script to UTXO’s.
- Removed the `height` from TariScript signatures and validation
- Update tari_crypto’s WASM interface to allow for clients to do private key arithmetic
- Update the grpc client used by the cucumber tests to support the new fields in the transaction protobuf
- Fixed a bug that causes panics when displaying an TransactionOutput with a malformed range proof
- Remove a number of duplicate Cucumber step file entries from a bad merge.

philipr-za committed:
Merge `Development` into the `TariScript` branch

This PR merges Development into the TariScript branch. The PR includes a few fixes to get the unit tests to pass.

sdbondi committed:
Remove "end of time" panic in emission

Removed `Option::expect` when total supply overflows `u64`. In this
case, the `EmissionRate` iterator considers itself "finished" and so
returns None as per the Iterator contract.

Removes the `Emission` trait, although nothing incorrect about using a
trait in this case, it did not provide much of an advantage (decoupling)
over using the concrete type. It also hid the Iterator impl, that should
have been used in `get_tokens_in_circulation`

Commit:
Merge #2880: [common] Fix Cucumber tests in `TariScript` branch

Fix Cucumber tests in `TariScript` branch

Commit:
Merge #2882: [common] Merge `Development` into the `TariScript` branch

Merge `Development` into the `TariScript` branch

SWvheerden committed:
New gen block for tari_script

Commit:
Consolidate config files into one

- Consolidated all the three configuration files that were
  in the code base into one file, usable by all operating
  systems.
- Removed some deprecated entries in the config file.
- Updated the installation files/installers to use the new
  config file.
- Clippy fixes.

sdbondi committed:
Adds version and commmit string to all application logs

- Adds the application name, version obtained form the manifest and
  commit sha to the logs on application startup
- Updates all log configs to route tari::application messages to all logs
- Moves all `tari_common` build-time dependencies and code behind `build` feature flag

Commit:
Add Bulletproofs Rewinding RFC

philipr-za committed:
[Wallet] Derive Comms Secret Key from the Master Seed key

This PR implements the derivation of the Comms private key in the wallet from the Master Seed Key that is used as the seed for the key managers used for the UTXO’s on the wall.

Previously the Comms Secret Key was generated by the client applications and it was stored in the Wallet Database. This key was also used to determine if the database was encrypted and if the provided passphrase is correct or not by storing the public key in the database and seeing if the stored private key corresponds to the permitted public key after decryption.

This PR removes the Comms Secret and Public keys from the wallet database and replaces them with the Master Seed Key which becomes the single secret key from which the Comms Secret Key, UTXO Spending Keys and Rewind keys are derived. The Comms Secret key is now derived when the wallet is started based on the persisted Master Secret Key. On the first run the wallet will generate the Master Secret Key and store it. The Output Manager Service used to generate and store this key but now it will accept it from the Wallet and use it to initialise its Key Managers. 

The Console Wallet and Wallet FFI is updated to account for this change.

SWvheerden committed:
Add one sided payment scanning and recovering

Refactor recovery task into service
change to hash based last scanned height
remove clippy errors

Commit:
Merge #2887: [common] New gen block for tari_script

New gen block for tari_script

SWvheerden committed:
update flaky unit tests

Add tx event to help with awaits

philipr-za committed:
FFI method removal and Test Fixes

Second commit contains a number of test fixes and some updates based on feedback from @kukabi about how the clients managed the Comms Secret Key. Because this is now completely handled in the backed we can remove the `comms_config_set_secret_key` function. It was also confirmed that the backend is fully responsible for managing the TorID so the old code that provided some migration support for the client to manage the TorID in the past has been removed.

Commit:
Verified massive multiple reorg cucumber

Commit:
Merge #2872: [wallet] Add one sided payment scanning and recovering

Add one sided payment scanning and recovering

Commit:
Merge #2886: [common] Remove "end of time" panic in emission (#2886)

Remove "end of time" panic in emission

Commit:
Merge branch 'tari-script' into general-app-version-in-logs

Commit:
Merge #2889: [common] Adds version and commmit string to all application logs (#2889)

Adds version and commmit string to all application logs

Commit:
Merge #2906: [tests] Verified massive multiple reorg cucumber (#2906)

Verified massive multiple reorg cucumber

Commit:
Merge #2895: [docs] Add Bulletproofs Rewinding RFC

Add Bulletproofs Rewinding RFC

Commit:
Merge #2890: [common] Consolidate config files into one

Consolidate config files into one

Commit:
Fix `header-stats` command

Tari-script network reset omitted changing this stibbons reference.

Commit:
Merge branch 'tari-script' into philip-derive-comms-private-key

Commit:
Merge #2894: [wallet] Derive Comms Secret Key from the Master Seed key

This PR implements the derivation of the Comms private key in the wallet from
the Master Seed Key that is used as the seed for the key managers used for the
UTXO’s on the wall.

Previously the Comms Secret Key was generated by the client applications and it
was stored in the Wallet Database. This key was also used to determine if the
database was encrypted and if the provided passphrase is correct or not by
storing the public key in the database and seeing if the stored private key
corresponds to the permitted public key after decryption.

This PR removes the Comms Secret and Public keys from the wallet database and
replaces them with the Master Seed Key which becomes the single secret key from
which the Comms Secret Key, UTXO Spending Keys and Rewind keys are derived. The
Comms Secret key is now derived when the wallet is started based on the
persisted Master Secret Key. On the first run the wallet will generate the
Master Secret Key and store it. The Output Manager Service used to generate and
store this key but now it will accept it from the Wallet and use it to
initialise its Key Managers.

The Console Wallet and Wallet FFI is updated to account for this change.

Commit:
Fix compile error

sdbondi committed:
Fix build error in console wallet

Commit:
Merge #2910: [chore] Fix build error in console wallet

Fix build error in console wallet

Commit:
fmt and clippy

Commit:
Merge #2908: [base-node] Fix `header-stats` command

Tari-script network reset omitted changing this stibbons reference.

Commit:
[ci] Switch to ssh urls for tari-crypto dep

Commit:
Merge #2913: [ci] Switch to ssh urls for tari-crypto dep

For whatever reason circle ci was failing to checkout the tari-crypto git repo,
switching these urls to use the ssh:// version is working

delta1 commented:
Just making a note here that it would be great to have a new tari-crypto release to crates.io for our dependencies

philipr-za committed:
Replace custom ChaCha20 cipher with RustCrypto ChaCha20 implementation

Fuzzing revealed that the custom ChaCha20 implementation in `tari_utilities` was not correct so this PR migrates the DHT encryption to use the audited RustCrypto ChaCha20 encryption. 

Had to bump the toolchain version for this update so updated to the most current version of nightly. The only update needed was adding a new feature flag #![feature(min_type_alias_impl_trait)] which is a new flag for the same issue we already use (i.e. #![feature(type_alias_impl_trait)])

Commit:
Merge #2897: [common] Replace custom ChaCha20 cipher

Replace custom ChaCha20 cipher with RustCrypto ChaCha20 implementation Fuzzing
revealed that the custom ChaCha20 implementation in tari_utilities was not
correct so this PR migrates the DHT encryption to use the audited RustCrypto
ChaCha20 encryption.

Had to bump the toolchain version for this update so updated to the most
current version of nightly. The only update needed was adding a new feature
flag #![feature(min_type_alias_impl_trait)] which is a new flag for the same
issue we already use (i.e. #![feature(type_alias_impl_trait)])

The new nightly toolchain contained an update to the linting used by rustfmt so
I have included a second commit that is just new formatting to make this easier
to review.

Commit:
Fix compiler warnings

Commit:
Merge branch 'development' into mb-fix-tari-script-branch

# Conflicts:
#	Cargo.lock
#	RFC/src/RFC-0201_TariScript.md
#	applications/tari_app_grpc/Cargo.toml
#	applications/tari_app_utilities/Cargo.toml
#	applications/tari_base_node/Cargo.toml
#	applications/tari_base_node/src/command_handler.rs
#	applications/tari_console_wallet/Cargo.toml
#	applications/tari_console_wallet/src/automation/commands.rs
#	applications/tari_console_wallet/src/main.rs
#	applications/tari_console_wallet/src/ui/components/send_tab.rs
#	applications/tari_console_wallet/src/wallet_modes.rs
#	applications/tari_merge_mining_proxy/Cargo.toml
#	applications/tari_mining_node/Cargo.toml
#	applications/test_faucet/Cargo.toml
#	base_layer/common_types/Cargo.toml
#	base_layer/core/Cargo.toml
#	base_layer/core/src/base_node/comms_interface/local_interface.rs
#	base_layer/core/src/base_node/state_machine_service/states/horizon_state_sync/horizon_state_synchronization.rs
#	base_layer/core/src/base_node/sync/rpc/service.rs
#	base_layer/core/src/blocks/genesis_block.rs
#	base_layer/core/src/chain_storage/accumulated_data.rs
#	base_layer/core/src/chain_storage/block_add_result.rs
#	base_layer/core/src/chain_storage/blockchain_database.rs
#	base_layer/core/src/chain_storage/historical_block.rs
#	base_layer/core/src/chain_storage/lmdb_db/lmdb_db.rs
#	base_layer/core/src/consensus/consensus_manager.rs
#	base_layer/core/src/consensus/emission.rs
#	base_layer/core/src/lib.rs
#	base_layer/core/src/mempool/service/inbound_handlers.rs
#	base_layer/core/src/test_helpers/blockchain.rs
#	base_layer/core/src/test_helpers/mod.rs
#	base_layer/core/src/validation/header_validator.rs
#	base_layer/core/src/validation/helpers.rs
#	base_layer/core/src/validation/mocks.rs
#	base_layer/core/tests/chain_storage_tests/chain_storage.rs
#	base_layer/core/tests/helpers/block_builders.rs
#	base_layer/key_manager/Cargo.toml
#	base_layer/mmr/Cargo.toml
#	base_layer/p2p/Cargo.toml
#	base_layer/service_framework/Cargo.toml
#	base_layer/wallet/Cargo.toml
#	base_layer/wallet/src/base_node_service/service.rs
#	base_layer/wallet/src/output_manager_service/service.rs
#	base_layer/wallet/src/storage/sqlite_db.rs
#	base_layer/wallet/src/tasks/wallet_recovery.rs
#	base_layer/wallet/src/transaction_service/service.rs
#	base_layer/wallet/src/transaction_service/storage/database.rs
#	base_layer/wallet/src/transaction_service/storage/sqlite_db.rs
#	base_layer/wallet/src/wallet.rs
#	base_layer/wallet/tests/wallet/mod.rs
#	base_layer/wallet_ffi/src/lib.rs
#	common/Cargo.toml
#	common/config/presets/tari_config_example.toml
#	comms/Cargo.toml
#	comms/dht/Cargo.toml
#	comms/rpc_macros/Cargo.toml
#	infrastructure/derive/Cargo.toml
#	infrastructure/shutdown/Cargo.toml
#	infrastructure/storage/Cargo.toml
#	infrastructure/test_utils/Cargo.toml
#	integration_tests/features/Reorgs.feature
#	integration_tests/features/WalletTransfer.feature
#	integration_tests/features/support/steps.js
#	integration_tests/helpers/transactionBuilder.js
#	integration_tests/helpers/util.js
#	integration_tests/package-lock.json

Commit:
Merge Dev into TariScript

SWvheerden committed:
Update and fix Import utxo functions


code review

SWvheerden committed:
rename to maturity

Commit:
Merge #2901:[wallet] Update and fix import utxo functions

Import utxo fn use defaults down the line, this will cause any UTXO that's
inserted into the wallet to be invalidated due to mismatched scripts, features,
etc.

Requires #2872

Commit:
Merge branch 'tari-script' into mb-fix-tari-script-branch

# Conflicts:
#	base_layer/wallet/src/transaction_service/service.rs
#	base_layer/wallet/src/wallet.rs
#	base_layer/wallet/tests/wallet/mod.rs
#	base_layer/wallet_ffi/src/lib.rs

Commit:
Merge Dev into TariScript

Commit:
Fix clippy

Commit:
Merge #2916: [common] Merge dev into tari-script

Add clippy warnings for cognitive complexity
libwallet-0.16.27
Fix re-org test
[base-node] Optimise pruned UTXO sync streaming protocol
[Wallet] Add mined height to display of mined transactions in Console Wallet
clarify one sided payment
[base-node] Fixes and tidies up prune mode cleanup
update tests to use sha3
add note
Add option to write unblinded UTXOs to CSV file
fmt
Test to reproduce target difficulty problem
fix case of cucumber
Insert pruned output audit
Fix incorrect boolean condition
v0.8.9
Fix potential overflow and improve performance of n_leaves
Add transaction washer util script
[base-node] Refactor chain storage and rule out some target diff bugs
[wallet] Update connectivity status when base node changes
Emission math audit
Consolidate config files into one
[wallet] Fixes possible duration overflow in crossterm loop
fix wallet recovery of coinbase
Fix cucumber test
Update Base Node Service monitor to shutdown correctly
[tests] eslint config and husky pre-commit hook
v0.8.10
Let wallet display pretty error for tor offline
Re-implemented methods for block explorer
[tests] Repair some integration tests
Verified massive multiple reorg cucumber
Bump timeouts
Remove unused search stxo
Remove unused search stxo
Fix tests and add results to CI
Fix compiler warnings
Merge Dev into TariScript
Merge Dev into TariScript
Fix clippy

Commit:
Merge branch 'development' into tari-script

# Conflicts:
#	Cargo.lock
#	applications/tari_app_utilities/Cargo.toml
#	applications/tari_base_node/src/recovery.rs
#	applications/tari_merge_mining_proxy/src/main.rs
#	applications/tari_merge_mining_proxy/src/test.rs
#	base_layer/core/src/base_node/comms_interface/inbound_handlers.rs
#	base_layer/core/src/chain_storage/blockchain_database.rs
#	base_layer/core/src/test_helpers/blockchain.rs
#	base_layer/core/src/test_helpers/mod.rs
#	base_layer/core/src/validation/header_validator.rs
#	base_layer/core/tests/block_validation.rs
#	base_layer/core/tests/helpers/block_builders.rs

Commit:
Merge dev

Commit:
Merge #2964: [chore] Merge dev v0.8.11 into tari-script

[MMProxy] Connection Limiting
Fix target difficulty
v0.8.11 Changes since v0.8.10
Merge dev

Commit:
Merge #2899: [tests] update flaky unit tests

update flaky unit tests

Commit:
Fix clippy warnings in tests

Commit:
chore: Fix clippy warnings in tests (#2965)

Commit:
fix: Update failing rust tests (#2961)
